crates/ark/src/modules/positron/graphics.R

```diff
@@ -134,107 +90,178 @@ renderWithPlotDevice <- function(filepath, format, width, height, res, type) {
-    })
+    # Store handle to current device (i.e. us)
+    old_dev <- grDevices::dev.cur()
 
-    # Turn off the graphics device.
-    grDevices::dev.off()
+    args <- finalizeDeviceArguments(format, width, height, pixel_ratio)
+    width <- args$width
+    height <- args$height
+    res <- args$res
+    type <- args$type
 
-    # Return path to the generated file.
-    invisible(filepath)
+    # Create a new graphics device.
+    switch(
+        format,
+        "png" = grDevices::png(
+            filename = path,
+            width = width,
+            height = height,
+            res = res,
+            type = type
+        ),
+        "svg" = grDevices::svg(
+            filename = path,
+            width = width,
+            height = height,
+        ),
+        "pdf" = grDevices::pdf(
+            file = path,
+            width = width,
+            height = height
+        ),
+        "jpeg" = grDevices::jpeg(
+            filename = path,
+            width = width,
+            height = height,
+            res = res,
+            type = type
+        ),
+        "tiff" = grDevices::tiff(
+            filename = path,
+            width = width,
+            height = height,
+            res = res,
+            type = type
+        ),
+        stop("Internal error: Unknown plot `format`.")
+    )
+
+    # Ensure we turn off the device on the way out, this:
+    # - Commits the plot to disk
+    # - Resets us back as being the current device
+    on.exit(utils::capture.output({
+        grDevices::dev.off()
+        if (old_dev > 1) {
+            grDevices::dev.set(old_dev)
+        }
+    }))
+
+    expr
+}
+
+finalizeDeviceArguments <- function(format, width, height, pixel_ratio) {
+    if (format == "png" || format == "jpeg" || format == "tiff") {
+        # These devices require `width` and `height` in pixels.
+        # We already have them in pixels, so we just scale by the `pixel_ratio`.
+        # `res` is nominal resolution specified in pixels-per-inch (ppi).
+        return(list(
+            type = defaultDeviceType(),
+            res = defaultResolutionInPixelsPerInch() * pixel_ratio,
+            width = width * pixel_ratio,
+            height = height * pixel_ratio
+        ))
+    }
+
+    if (format == "svg" || format == "pdf") {
+        # These devices require `width` and `height` in inches.
+        # We convert from pixels to inches here.
+        # There is no `type` or `res` argument for these devices.
+        return(list(
+            type = NULL,
+            res = NULL,
+            width = width * pixel_ratio / defaultResolutionInPixelsPerInch(),
+            height = height * pixel_ratio / defaultResolutionInPixelsPerInch()
+        ))
+    }
+
+    stop("Internal error: Unknown plot `format`.")
+}
+
+defaultResolutionInPixelsPerInch <- function() {
+    if (Sys.info()[["sysname"]] == "Darwin") {
+        96L
+    } else {
+        72L
+    }
+}
```

Comment:
These were here before, but are they right?

Reply:
They are more correct than any other numbers but that doesn't mean they are truly correct. macOS uses 96 DPI as it's internal scaling, whereas this is user definable on windows. However, this is a scaling factor that tries to make things that appear on screen be as close to the size they are printed in and this has always been a perilous proposition